### ISSUE TITLE
Support s3:// URIs in "source" field for aws_s3_bucket_object resource

### ISSUE THREAD

**Issue by lorengordon** [hashicorp/terraform-provider-aws#3943]
I was hoping to be able to use Terraform to control a pipeline of S3 objects at different stages, basically just copying an object from one bucket to another. I thought I'd be able to specify an `aws_s3_bucket_object` resource, using an S3 URI in the `source` field but right now it just errors.

### Terraform Version
`terraform -v`:
```
Terraform v0.11.5
+ provider.aws v1.12.0
```

### Affected Resource(s)
Please list the resources as a list, for example:
- aws_s3_bucket_object

If this issue appears to affect multiple resources, it may be an issue with Terraform's core, so please mention this.

### Terraform Configuration Files
```hcl
resource "aws_s3_bucket" "foo" {
  bucket = "foo-123321-test"
}

resource "aws_s3_bucket" "bar" {
  bucket = "bar-123321-test"
}

resource "aws_s3_bucket_object" "foo" {
  bucket = "${aws_s3_bucket.foo.id}"
  key    = "foo"
  content = "bar"
}

resource "aws_s3_bucket_object" "bar" {
  bucket = "${aws_s3_bucket.bar.id}"
  key    = "bar"
  source = "s3://${aws_s3_bucket.foo.id}/${aws_s3_bucket_object.foo.id}"
}
```

### Debug Output

I can generate debug output if you really want, but I'm pretty sure that S3 URIs just are not currently supported in the `source` field.

```
Error: Error applying plan:

1 error(s) occurred:

* aws_s3_bucket_object.bar: 1 error(s) occurred:

* aws_s3_bucket_object.bar: Error opening S3 bucket object source (s3://foo-123321-test/foo): open s3://foo-123321-test/foo: The filename, directory name, or volume label syntax is incorrect.
```

### Expected Behavior

The file specified by the S3 URI should have been copied to the destination bucket/key. I'd expect using  [CopyObject](https://docs.aws.amazon.com/sdk-for-go/api/service/s3/#S3.CopyObject).

### Actual Behavior

Threw an error.

### Steps to Reproduce

1. `terraform apply`

**Comment by lorengordon** [comment]
Please reopen. We have an open pr in to provide the capability. Just need someone to review and merge.

**Comment by lorengordon** [comment]
Implemented in https://github.com/hashicorp/terraform-provider-aws/pull/15461, [released in v3.29.0](https://github.com/hashicorp/terraform-provider-aws/blob/v3.29.0/CHANGELOG.md?rgh-link-date=2021-02-19T16%3A55%3A32Z).
